lib/commands/dialog/index.js

```diff
@@ -68,51 +68,58 @@ class DialogCommand extends AbstractCommand {
      * @return { skillId, locale, stage, profile, debug, replayFile, smapiClient, userInputs <for replay mode> }
      */
-    _getDialogConfig(cmd) {
+    _getDialogConfig(cmd, callback) {
         const debug = Boolean(cmd.debug);
         let { skillId, locale, stage, profile } = cmd;
         profile = profileHelper.runtimeProfile(profile);
         stage = stage || CONSTANTS.SKILL.STAGE.DEVELOPMENT;
         let firstLocaleFromManifest;
         let userInputs;
+        const smapiClient = new SmapiClient({ profile, doDebug: debug });
         if (cmd.replay) {
             const dialogReplayConfig = new DialogReplayFile(cmd.replay);
             skillId = dialogReplayConfig.getSkillId();
             if (!stringUtils.isNonBlankString(skillId)) {
-                throw 'Replay file must contain skillId';
+                return callback(new CliError('Replay file must contain skillId'));
             }
             locale = dialogReplayConfig.getLocale();
             if (!stringUtils.isNonBlankString(locale)) {
-                throw 'Replay file must contain locale';
+                return callback(new CliError('Replay file must contain locale'));
+            }
+            try {
+                userInputs = this._validateUserInputs(dialogReplayConfig.getUserInput());
+            } catch (err) {
+                return callback(err);
             }
-            userInputs = this._validateUserInputs(dialogReplayConfig.getUserInput());
         } else {
             if (!stringUtils.isNonBlankString(skillId)) {
                 try {
                     new ResourcesConfig(path.join(process.cwd(), CONSTANTS.FILE_PATH.ASK_RESOURCES_JSON_CONFIG));
                     skillId = ResourcesConfig.getInstance().getSkillId(profile);
                 } catch (err) {
-                    throw new CliError('Failed to read project resource file. Please run the command within a ask-cli project.');
+                    return callback(new CliError('Failed to read project resource file. Please run the command within a ask-cli project.'));
                 }
                 if (!stringUtils.isNonBlankString(skillId)) {
-                    throw `Failed to obtain skill-id from project resource file ${CONSTANTS.FILE_PATH.ASK_RESOURCES_JSON_CONFIG}`;
+                    return callback(new CliError('Failed to obtain skill-id from project '
+                    + `resource file ${CONSTANTS.FILE_PATH.ASK_RESOURCES_JSON_CONFIG}`));
                 }
-                const skillPackageSrc = ResourcesConfig.getInstance().getSkillMetaSrc(profile);
-                const manifestPath = path.join(skillPackageSrc, CONSTANTS.FILE_PATH.SKILL_PACKAGE.MANIFEST);
-                new Manifest(manifestPath);
-                [firstLocaleFromManifest] = Object.keys(Manifest.getInstance().getPublishingLocales());
             }
-
+        }
+        smapiClient.skill.manifest.getManifest(skillId, stage, (err, res) => {
```

Comment:
can we somehow avoid calling this api again in the next step: helper.validateDialogArgs? I guess validateDialogArgs doesn't always need dialogMode as the input.

Reply:
we can move logic of validateDialogArgs to _getDialogConfig. _getDialogConfig does a lot of validation already.